File: api/src/com/proyecto404/backoffice/modules/security/SecurityModule.kt
package com.proyecto404.backoffice.modules.security

import com.proyecto404.backoffice.modules.common.Core
import com.proyecto404.backoffice.modules.security.app.CreateUser
import com.proyecto404.backoffice.modules.security.app.Login
import com.proyecto404.backoffice.modules.security.domain.sessionToken.UUIDSessionTokenFactory
import com.proyecto404.backoffice.modules.security.infrastructure.middlewares.AuthenticationMiddleware
import com.proyecto404.backoffice.modules.security.infrastructure.middlewares.AuthorizationMiddleware
import com.proyecto404.backoffice.modules.security.infrastructure.passwords.RandomSaltGenerator
import com.proyecto404.backoffice.modules.security.infrastructure.passwords.SHA512PasswordEncryptor

class SecurityModule(coreConfig: Core.Config, coreServices: Core.Services) {
    private val cqBus = coreServices.cqBus
    private val repositories = coreServices.repositories
    private val saltGenerator = RandomSaltGenerator()
    private val passwordEncryptor = SHA512PasswordEncryptor()
    private val sessionTokenFactory = UUIDSessionTokenFactory()

    init {
        with(cqBus) {
            registerMiddleware(AuthorizationMiddleware())
            registerMiddleware(AuthenticationMiddleware(repositories.get()))

            registerHandler { CreateUser.Handler(repositories, saltGenerator, passwordEncryptor) }
            registerHandler { Login.Handler(repositories, passwordEncryptor, sessionTokenFactory) }
        }
    }
}

File: webapp/src/core/security/SecurityModule.ts
import { Login } from './app/Login'
import { HttpLoginHandler } from './infrastructure/http/HttpLoginHandler'
import { CoreConfig, Services } from '../Core'
import {CreateContext} from "./app/GetContext"
import {HttpCreateContextHandler} from "./infrastructure/http/HttpCreateContextHandler"

export class SecurityModule {
    constructor(private config: CoreConfig, private services: Services) {
        const cqBus = config.cqBus
        cqBus.registerHandler(Login, () => new HttpLoginHandler(services.config.httpClient))
        cqBus.registerHandler(CreateContext, () => new HttpCreateContextHandler(services.config.httpClient))
    }
}
